src/lib.rs

```diff
@@ -383,57 +478,104 @@ impl<T> Receiver<T> {
                 std::thread::sleep(std::time::Duration::from_millis(10));
 
                 // Write our waker instance to the channel.
+                // SAFETY: we are still in the empty state, meaning that the sender will not try
+                // to access the waker until it sees the state set to RECEIVING
                 unsafe { channel.write_waker(ReceiverWaker::current_thread()) };
 
-                match channel
-                    .state
-                    .compare_exchange(EMPTY, RECEIVING, SeqCst, SeqCst)
-                {
+                // Switch the state from EMPTY to RECEIVING. We need to do this in one
```

Comment:
Same as above. We don't really know that we are in the `EMPTY` state. I think it would be clearer to state that we want to set the state to `RECEIVING` while also fetching the previous state to act on it.